launch/bash_test.go

```diff
@@ -0,0 +1,209 @@
+		tmpDir string
+	)
+
+	it.Before(func() {
+		h.SkipIf(t, runtime.GOOS == "windows", "skip bash tests on windows")
+		var err error
+		tmpDir, err = ioutil.TempDir("", "shell-test")
+		h.AssertNil(t, err)
+		shell = &launch.BashShell{Exec: hl.SyscallExecWithStdout(t, tmpDir)}
+	})
+
+	it.After(func() {
+		h.AssertNil(t, os.RemoveAll(tmpDir))
+	})
+
+	when("#Launch", func() {
+		var process launch.ShellProcess
+
+		when("script", func() {
+			when("there are profiles", func() {
+				it.Before(func() {
+					process = launch.ShellProcess{
+						Script:  true,
+						Command: `printf "profile env: '%s'" "$PROFILE_VAR"`,
+						Caller:  "some-profile-argv0",
+						Env: []string{
+							"SOME_VAR=some-val",
+						},
+					}
+					process.Profiles = []string{
+						filepath.Join("testdata", "profiles", "print_argv0"),
+						filepath.Join("testdata", "profiles", "print_env"),
+						filepath.Join("testdata", "profiles", "set_env"),
+					}
+				})
+
+				it("sets argv0 for profile scripts", func() {
+					err := shell.Launch(process)
+					h.AssertNil(t, err)
+					stdout := rdfile(t, filepath.Join(tmpDir, "stdout"))
+					if len(stdout) == 0 {
+						stderr := rdfile(t, filepath.Join(tmpDir, "stderr"))
+						t.Fatalf("stdout was empty: stderr: %s\n", stderr)
+					}
+					h.AssertStringContains(t, stdout, "profile argv0: 'some-profile-argv0'")
+				})
+
+				it("sets env for profile scripts", func() {
+					err := shell.Launch(process)
+					h.AssertNil(t, err)
+					stdout := rdfile(t, filepath.Join(tmpDir, "stdout"))
+					if len(stdout) == 0 {
+						stderr := rdfile(t, filepath.Join(tmpDir, "stderr"))
+						t.Fatalf("stdout was empty: stderr: %s\n", stderr)
+					}
+					h.AssertStringContains(t, stdout, "SOME_VAR: 'some-val'")
+				})
+
+				it("env vars set in profile scripts are available to the command", func() {
+					err := shell.Launch(process)
+					h.AssertNil(t, err)
+					stdout := rdfile(t, filepath.Join(tmpDir, "stdout"))
+					if len(stdout) == 0 {
+						stderr := rdfile(t, filepath.Join(tmpDir, "stderr"))
+						t.Fatalf("stdout was empty: stderr: %s\n", stderr)
+					}
+					h.AssertStringContains(t, stdout, "profile env: 'some-profile-var'")
+				})
+			})
+
+			it("sets env", func() {
+				process = launch.ShellProcess{
+					Script:  true,
+					Command: `printf "SOME_VAR: '%s'" "$SOME_VAR"`,
+					Caller:  "some-profile-argv0",
+					Env: []string{
+						"SOME_VAR=some-val",
+					},
+				}
+				err := shell.Launch(process)
+				h.AssertNil(t, err)
+				stdout := rdfile(t, filepath.Join(tmpDir, "stdout"))
+				if len(stdout) == 0 {
+					stderr := rdfile(t, filepath.Join(tmpDir, "stderr"))
+					t.Fatalf("stdout was empty: stderr: %s\n", stderr)
+				}
+				h.AssertStringContains(t, stdout, "SOME_VAR: 'some-val'")
+			})
+
+			it("provides args to bash", func() {
+				process = launch.ShellProcess{
+					Script:  true,
+					Command: `printf "SOME_ARG: '%s'" "$1"`,
+					Args:    []string{"", "some arg1"},
```

Comment:
Functionally it doesn't matter to this test we could have printed `$0` and dropped the empty first arg. But because `$0` is supposed to represent the process being invoked it feels weird to put an actual argument there. We could change `$0` to `bash` to make it more realistic/readable.